storage/innobase/trx/trx0purge.cc

```diff
@@ -356,6 +357,8 @@ inline dberr_t purge_sys_t::iterator::free_history_rseg(trx_rseg_t &rseg) const
   mtr_t mtr;
   bool freed= false;
   uint32_t rseg_ref= 0;
+  const auto last_boffset= srv_page_size - TRX_UNDO_LOG_OLD_HDR_SIZE;
+  const auto last_page= rseg.space->free_limit;
 
```

Comment:
Extending a tablespace should be covered by an exclusive `rseg.space->latch`, which we are actually not holding here. We are only holding an exclusive `rseg.latch`. With `innodb_undo_tablespaces=0` the tablespace could in fact be extended concurrently, and there could be a race condition around reading the `free_limit`. I think that it is harmless; on no CPU architecture that I can think of would we read a complete garbage value. Even if the tablespace was being extended or `free_limit` increased concurrently, using a previous tighter limit does not matter, because the pre existing pages in the list must be within the old `free_limit`. Yes, other pages in the tablespace can be written, but no new pages can be added to this rollback segment. I will add a source code comment about this.